core/arch/arm/kernel/user_ta.c

```diff
@@ -824,6 +826,130 @@ static TEE_Result set_seg_prot(struct user_ta_ctx *utc,
 	return res;
 }
 
+#ifdef CFG_UNWIND
+
+/* An item in the exception index table */
+struct unwind_idx {
+	uint32_t offset;
+	uint32_t insn;
+};
+
+#define	EXIDX_CANTUNWIND 1
+
+static uint32_t offset_prel31(uint32_t addr, int32_t offset)
+{
+	return (addr + offset) & 0x7FFFFFFFUL;
+}
+
+static TEE_Result relocate_exidx(void *exidx, size_t exidx_sz, int32_t offset)
+{
+	size_t num_items = exidx_sz / sizeof(struct unwind_idx);
+	struct unwind_idx *start = (struct unwind_idx *)exidx;
+	size_t n;
+
+	start = (struct unwind_idx *)exidx;
+	for (n = 0; n < num_items; n++) {
+		struct unwind_idx *item = &start[n];
+
+		if (item->offset & BIT32(31))
+			return TEE_ERROR_BAD_FORMAT;
+
+		/* offset to the start of the function has to be adjusted */
+		item->offset = offset_prel31(item->offset, offset);
+
+		if (item->insn == EXIDX_CANTUNWIND)
+			continue;
+		if (item->insn & BIT32(31)) {
+			/* insn is a table entry itself */
+			continue;
+		}
+		/*
+		 * insn is an offset to an entry in .ARM.extab so it has to be
+		 * adjusted
+		 */
+		item->insn = offset_prel31(item->insn, offset);
+	}
+	return TEE_SUCCESS;
+}
+
+/*
+ * 32-bit TAs: set the address and size of the exception index table (EXIDX).
+ * If the TA contains only one ELF, we point to its table. Otherwise, a
+ * consolidated table is made by concatenating the tables found in each ELF and
+ * adjusting their content to account for the offset relative to the original
+ * location.
+ */
+static TEE_Result set_exidx(struct user_ta_ctx *utc)
+{
+	struct user_ta_elf *exe;
+	struct user_ta_elf *elf;
+	vaddr_t exidx;
+	size_t exidx_sz = 0;
+	TEE_Result res;
+	uint8_t *p;
+
+	if (!utc->is_32bit)
+		return TEE_SUCCESS;
+
+	exe = TAILQ_FIRST(&utc->elfs);
+	if (!TAILQ_NEXT(exe, link)) {
+		/* We have a single ELF: simply reference its table */
+		utc->exidx_start = exe->exidx_start;
+		utc->exidx_size = exe->exidx_size;
+		return TEE_SUCCESS;
+	}
+
+	TAILQ_FOREACH(elf, &utc->elfs, link)
+		exidx_sz += elf->exidx_size;
+
+	utc->mobj_exidx = alloc_ta_mem(exidx_sz);
+	if (!utc->mobj_exidx)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	exidx = 0;
+	res = vm_map(utc, &exidx, exidx_sz, TEE_MATTR_UR | TEE_MATTR_PRW,
+		     utc->mobj_exidx, 0);
+	if (res)
+		goto err;
+	DMSG("New EXIDX table mapped at 0x%" PRIxVA " size %zu",
+	     exidx, exidx_sz);
+
+	p = (void *)exidx;
+	TAILQ_FOREACH(elf, &utc->elfs, link) {
+		void *e_exidx = (void *)(elf->exidx_start + elf->load_addr);
+		size_t e_exidx_sz = elf->exidx_size;
+		int32_t offs = (int32_t)((vaddr_t)e_exidx - (vaddr_t)p);
+
+		memcpy(p, e_exidx, e_exidx_sz);
+		res = relocate_exidx(p, e_exidx_sz, offs);
+		if (res)
+			goto err;
+		p += e_exidx_sz;
+	}
+
+	res = vm_set_prot(utc, exidx,
+			  ROUNDUP(exidx_sz, SMALL_PAGE_SIZE),
+			  TEE_MATTR_UR);
```

Comment:
why adding here user read access, isn't it already  done above?

Reply:
This is actually removing kernel RW. Now that you're pointing this, I'm wondering why kernel read permission is not needed, since the stack unwinding is done in kernel mode?

Reply:
Ok, i understand. thanks.

Reply:
I guess that for correctness we should have `| TEE_MATTR_PR`. However, in practice it doesn't matter as privileged read always is enabled if unprivileged read is enabled.

Reply:
@jenswi-linaro now that makes sense. I'm adding a comment that I personally find useful, but if it's too obvious for the "person skilled in the art" then let me know ;-)